multicluster/service-mirror/cluster_watcher.go

```diff
@@ -968,28 +999,52 @@ func (rcsw *RemoteClusterServiceWatcher) repairEndpoints(ctx context.Context) er
 
 		endpoints, err := rcsw.localAPIClient.Endpoint().Lister().Endpoints(svc.Namespace).Get(svc.Name)
 		if err != nil {
-			rcsw.log.Errorf("Could not get local endpoints: %s", err)
-			continue
+			if !kerrors.IsNotFound(err) {
+				rcsw.log.Errorf("failed to list local endpoints: %s", err)
+				continue
+			}
+			endpoints, err = rcsw.localAPIClient.Client.CoreV1().Endpoints(svc.Namespace).Get(ctx, svc.Name, metav1.GetOptions{})
+			if err != nil {
+				rcsw.log.Errorf("failed to get local endpoints %s/%s: %s", svc.Namespace, svc.Name, err)
+			}
 		}
 
 		updatedEndpoints := endpoints.DeepCopy()
-		updatedEndpoints.Subsets = []corev1.EndpointSubset{
-			{
-				Addresses: gatewayAddresses,
-				Ports:     rcsw.getEndpointsPorts(updatedService),
-			},
-		}
+		if !rcsw.alive {
+			rcsw.log.Warnf("gateway for service %s/%s does not have ready addresses; updating endpoint subsets to not ready", updatedService.Namespace, updatedService.Name)
+
+			// The gateway is not alive so the Endpoints update should reflect
+			// that by changing the gateway addresses to not ready. This
+			// ensures that mirror services on the source cluster do not
+			// attempt to send any traffic to the target gateway.
+			updatedEndpoints.Subsets = []corev1.EndpointSubset{
+				{
+					NotReadyAddresses: gatewayAddresses,
+					Ports:             rcsw.getEndpointsPorts(updatedService),
+				},
+			}
+		} else {
+			rcsw.log.Debugf("gateway for service %s/%s has addresses; updating endpoint subsets to: %v", updatedService.Namespace, updatedService.Name, gatewayAddresses)
 
-		// If the Service's Endpoints has no Subsets, use an empty Subset locally as well
-		targetService := svc.DeepCopy()
-		targetService.Name = rcsw.targetResourceName(svc.Name)
-		empty, err := rcsw.isEmptyService(targetService)
-		if err != nil {
-			rcsw.log.Errorf("could not check service emptiness: %s", err)
-			continue
-		}
-		if empty {
-			updatedEndpoints.Subsets = []corev1.EndpointSubset{}
+			updatedEndpoints.Subsets = []corev1.EndpointSubset{
+				{
+					Addresses: gatewayAddresses,
+					Ports:     rcsw.getEndpointsPorts(updatedService),
+				},
+			}
+
+			// If the Service's Endpoints has no Subsets, use an empty Subset locally as well
+			targetService := svc.DeepCopy()
+			targetService.Name = rcsw.targetResourceName(svc.Name)
+			empty, err := rcsw.isEmptyService(targetService)
+			if err != nil {
+				rcsw.log.Errorf("could not check service emptiness: %s", err)
+				continue
+			}
+			if empty {
+				rcsw.log.Warnf("exported service %s/%s is empty", targetService.Namespace, targetService.Name)
+				updatedEndpoints.Subsets = []corev1.EndpointSubset{}
```

Comment:
Here you set all subsets to be empty whereas above (if the gateway is not alive) you set `NotReadyAddresses`. Is there a difference between these two ways of "emptying" Endpoints?

Reply:
I think this keeps the state of the source cluster closer to the target cluster.

In the case that the gateway is alive but an exported service has no endpoints, then on the source cluster the mirror cluster will also have no endpoints. I think this is the more important "state" we get from this.

In the case that the gateway is down, a mirror service is essentially blocked on gateway readiness so that's why we indicate its endpoints are not ready. Once the gateway is alive, we'll repair again and get closer to replicating the state; we'll get to this `else` block and either update the mirror service endpoints to ready, or completely empty them.